---
title: visualisation.py
---

This module takes in a taxonomy feature abundance table, the study metadata and performs the following comparisons between user-defined groups:

1. Alpha-diversity comparisons between groups where alpha-diversity is measured by a range of indices such as: Shannon, Simpson, Chao1, Chao2, etc.
2. Beta-diversity comparisons between groups where beta-diversity is measured using Bray-Curtis dissimilarity measure
3. Differential abundance feature comparisons between groups using univariate approach


***


# Basic usage

```
$ singularity run --app mima-visualisation $SANDBOX \
--feature-table <path/to/taxonomy_profiling/feature-abundance-table> \
--metadata <path/to/project/metadata.tsv> \
--study-groups <comma,separated,list> \
--output-dir <path/to/output> > visualisation.log
```

## Required inputs

| Parameter | Description |
|:----------|:------------|
| `--feature-table` | Path to taxonomy feature table (e.g. output from [Taxonomy profiling](taxonomy_profiling) module). The rows are features (taxon) and the columns are samples |
| `--metadata` | Path to study metadata file, where first column is the `Sample_ID` and must match with the columns in the feature table |
| `--study-group` | Comma separated list of column names in the `metadata` input table which contains the study groups of interest for comparison |
| `--output-dir` | Path to output directory where analysis results will be saved. By default, the analysis for alpha-diversity, beta-diversity and differential abundance analysis will be performed. See [Outputs](#outputs) for detailed description. |


## Outputs

Several outputs will be generated by this module, the files are organised into the following directory structure:

```
├── Kraken
├── Kraken_alpha
│   ├── boxplots
│   │   ├── corrected_sig
│   │   ├── non_sig
│   │   └── uncorrected_sig
│   └── wilcox-pairwise
├── Kraken_beta
│   ├── permanova
│   └── plots
└── Kraken_diff-abundance
    ├── barplot
    ├── boxplots
    │   ├── corrected_sig
    │   ├── non_sig
    │   └── uncorrected_sig
    ├── volcano
    └── wilcox-pairwise
```

| Output    | Description                                      |
|-----------|--------------------------------------------------|
|`Kraken`   | directory that contains two transposed feature tables where rows are samples and columns are features (`otu_matrix.txt`). The `TaxProfileWithMeta.txt` text file combines the feature abundance table concatenated with the metadata table. |
|`Kraken_alpha` | contains *alpha-diversity* comparison test results and plots. Boxplots are further divided into three sub-folders: `corrected_sig`, `non_sig` and `uncorrected_sig`. By default, no plots will be generated for insignificant features, this can be enabled by using the `--plot-insignificant` parameter setting (see full help below) |
|`Kraken_beta` | contains *beta-diversity* comparison test results and PCoA plots. |
|`Kraken_diff-abundance` | contains *differential-abundance* analysis results and plots. The boxplots are also further divided into three sub-folders: `corrected_sig`, `non_sig`, and `uncorrected_sig`. By default, no plots will be generated for insignificant features, this can be enabled by using the `--plot-insignificant` parameter setting (see full help below) |


***


# Full help

```
usage: visualisation.py --feature-table FEATURE_TABLE [FEATURE_TABLE ...] --metadata METADATA
                        --study-groups STUDY_GROUPS --output-dir OUTPUT_DIR
                        [--plot-insignificant] [--ini-file INI_FILE] [-h] [--verbose] [--debug]

Visualisation module part of the MRC Metagenomics pipeline

[1] Required arguments:
  --feature-table FEATURE_TABLE [FEATURE_TABLE ...]
                        Path to the tab-separated feature table file where rows are Features and
                        columns are Samples. The first column header should be either:
                        <empty>|ID|# Pathway|# Gene|# IGGsearch species table Other heading
                        columns are current unsupported and no output will be generated.
  --metadata METADATA   Path to the tab-/comma-separated metadata file
  --study-groups STUDY_GROUPS
                        Comma separated list of column names in the metadata table which contain
                        the study groups of interest
  --output-dir OUTPUT_DIR
                        Output directory

[2] Processing arguments:
  --plot-insignificant  turn on to generate plots for insignificant comparisons
  --ini-file INI_FILE   Configuration file for figure parameters e.g., size [default:
                        figures.ini]

[3] Optional arguments:
  -h, --help            show this message and exit
  --verbose             turn on verbose messages
  --debug               turn on debugging messages
```