---
title: Quality control (QC)
description: QC module in MIMA, check reads are of high standard
weight: 110
---


{{% pageinfo color=danger %}}
This step must be done before Taxonomy and Function profiling.
{{% /pageinfo %}}

## Introduction: QC module

Quality control (QC) checks the sequenced reads obtained from the sequencing machine is of good quality. Bad quality reads or artefacts due to sequencing error if not removed can lead to spurious results and affect downstream analyses. There are a number of tools available for checking the read quality of high-throughput sequencing data.

This pipeline uses the following tools:

* <a href="https://jgi.doe.gov/data-and-tools/software-tools/bbtools/" target="_blank">BBTool suite</a>
* <a href="https://github.com/OpenGene/fastp" target="_blank">FastP</a>
* <a href="https://github.com/lh3/minimap2" target="_blank">Minimap2</a>

### Workflow description

One bash (*.sh) script per sample is generated by this module that performs the following steps. The module also generates $N$ number of PBS scripts (set using the [`--num-pbs-jobs` parameter](#parameters-explained)). 

This tutorial has 9 samples spread across 3 PBS jobs. One PBS job will execute three bash scripts, one per sample.

<table class="table table-borderless">
<tr>
  <th>Step</th>
  <th></th>
</tr>
<tr>
  <td>1. <b>repair.sh</b> from BBTools/BBMap tool suite, repairs the sequenced reads and separates out any singleton reads (orphaned reads that are missing either the forward or reverse partner read) into its own text file.</td>
  <td rowspan=5 style="width:40%"><img src="../images/tut_QCpipeline.png"/></td>
</tr>
<tr><td>2. <b>dereplication</b> using clumpify.sh from BBTools/BBMap tool suite, removes duplicate reads and clusters reads for faster downstream processing</td></tr>
<tr><td>3. <b>quality checking</b> is done with fastp.sh and checks the quality of the reads and removes any reads that are of low quality, too long or too short</td></tr>
<tr><td>4. <b>decontamination</b> with minimap2.sh maps the sequenced reads against a reference genome (e.g., human) defined by the user. It removes these host-reads from the data and generates a <i>cleaned</i> output file. This becomes the input for taxonomy profiling and function profiling modules.</td></tr>
<tr><td>The (optional) last step generates a QC report text file with summary statistics for each file. This step is run <b>after all</b> PBS scripts have been executed for all samples in the study.</td></tr>
</table>

For details about the tool versions, you can [check the MIMA container image installed]({{< ref "installation.md#confirm-installation" >}})


## Step 1. Generate QC PBS scripts

- Tick off the [check list]({{< ref "data-processing#check-list" >}})
- Find the [absolute path]({{< ref "need-to-know.md#use-absolute-paths" >}}) to the host reference genome file (e.g., Human host might use GRCh38_latest_genome.fna)
- Replace the highlighted line `--ref <path/to/human/GRCh38_latest_genome.fna>` to the location of your host reference genome file
- Enter the following command 
  - the backslash (`\`) at the end of each line informs the terminal that the command has not finished and there's more to come (we broke up the command for readability purposes to explain each parameter below)
  - *note* if you enter the command as one line, remove the backslashes


{{< highlight Tcsh "linenos=table,hl_lines=6,linenostart=1" >}}
apptainer run --app mima-qc $SANDBOX \
-i ~/mima_tutorial/raw_data \
-o ~/mima_tutorial/output \
-m ~/mima_tutorial/manifest.csv \
--num-pbs-jobs 3 \
--ref <path/to/human/GRCh38_latest_genome.fna> \
--mode container \
--pbs-config ~/mima_tutorial/pbs_header_qc.cfg
{{< /highlight >}}


{{% alert info=warn %}}
For MRC users, see <a href="https://unsw.sharepoint.com/:w:/r/sites/mrc_bioinformatics_unit/_layouts/15/doc2.aspx?sourcedoc=%7BBF4C59B2-9C8D-4868-AB8C-5B6B4C206252%7D&file=Katana%20reference%20databases%20-%20MIMA%20pipeline.docx&action=default&mobileredirect=true" target="_blank;">here for file locations</a>
{{% /alert %}}


### Parameters explained

| <div style="width:150px">Parameter</div> | <div style="width:150px">Required?</div> | Description |
| ---------- | ---------| ----------- |
| `-i <input>` | yes | must be [absolute path]({{< ref "need-to-know.md#use-absolute-paths" >}}) to the directory where fastq files are stored. FastQ files often have the \*.fastq.gz or \*.fq.gz extension. This path is used to find the *FileID\_R1* and *FileID\_R2* columns specified in the `manifest.csv` file |
| `-o <output>` | yes | <p>absolute path output directory where results will be saved. The `<output>` path will be created if it does not exists.</p><p><small>**Note** if there are already existing subdirectories in the `<output>` path, then this step will fail.</small></p> |
| `-m <manifest.csv>` | yes | <p>a comma-separated file (\*.csv) that has three columns with the headers: **Sample\_ID, FileID\_R1, FileID\_R2** see the example below.</p><p><small>**Note** the fileID\_R1 and fileID\_R2 are relative to the `-i <input>` path provided.</small></p> |
| `--num-pbs-jobs` | no<br/><small>(default=4)</small> | number of PBS scripts to generate by default 4 jobs are created with samples split equally between the jobs |
| `--ref` | yes | path to the host genome file GRCh38_latest_genome.fna |
| `--mode container` | no<br/><small>(default='single')</small> | set this if you are running as a Container |
| `--pbs-config` | yes if<br/> `--mode container` | path to the pbs configuration file, must specify this parameter if `--mode container` is set | 

{{% alert color=info title=tip %}}
These commands can get really long, you can use Notepad to

- copy \& paste the commands
- change the paths and parameters as desired
- copy \& paste to the terminal

Or you can explore using the <a href="https://www.vim.org/" target="_blank;">vim</a> text editor.
{{% /alert %}}

## Step 2. Check generated scripts

* After step 1, you should see the new directory: `~/mima_tutorial/output`
* Inside `output/`, there should be a sub-directory `QC_module`
* Inside `output/QC_module`, there should be 3 PBS scripts with the \*.pbs file extension

```Shell
tree ~/mima_tutorial
```

- only the `output` folder is shown in the snapshot below

```Text
~/mima_tutorial
└── output/
    └──  QC_module
      ├── CleanReads
      ├── qcModule_0.pbs
      ├── qcModule_1.pbs
      ├── qcModule_2.pbs
      ├── QCReport
      ├── SRR17380115.sh
      ├── SRR17380118.sh
      ├── SRR17380122.sh
      ├── SRR17380209.sh
      ├── SRR17380218.sh
      ├── SRR17380222.sh
      ├── SRR17380231.sh
      ├── SRR17380232.sh
      └── SRR17380236.sh
```

## Step 3. Submit QC jobs

- Let's examine one of the PBS scripts to be submitted

```Shell
cat ~/mima_tutorial/output/QC_module/qcModule_0.pbs
```

* Your PBS script should look something like below, with some differences
  - line 10: `IMAGE_DIR` should be where you installed MIMA and [build the sandbox]({{< ref " installation.md#build-a-sandbox" >}})
  - line 11: `APPTAINER_BIND` should be setup during installation when [binding paths]({{< ref "what-is-container.md#path-binding" >}})
    - make sure to include the path where the host reference genome file is located
  - line 14: `/home/user` is replaced with the [absolute path]({{< ref "need-to-know.md#use-absolute-paths" >}}) to your actual home directory

{{< highlight Bash "linenos=table,hl_lines=10-11 14,linenostart=1" >}}
#!/bin/bash
#PBS -N QC_module_0
#PBS -l ncpus=8
#PBS -l walltime=2:00:00
#PBS -l mem=64GB
#PBS -j oe

set -x

IMAGE_DIR=~/mima-pipeline
export APPTAINER_BIND="/path/to/human/GRCh38:/path/to/human/GRCh38"


cd /home/user/mima_tutorial/output/QC_module/

singularity exec ${IMAGE_DIR} bash /home/user/mima_tutorial/output/QC_module/SRR17380209.sh > SRR17380209_qc_module.log 2>&1
singularity exec ${IMAGE_DIR} bash /home/user/mima_tutorial/output/QC_module/SRR17380232.sh > SRR17380232_qc_module.log 2>&1
singularity exec ${IMAGE_DIR} bash /home/user/mima_tutorial/output/QC_module/SRR17380236.sh > SRR17380236_qc_module.log 2>&1
{{< /highlight >}}


- Navigate to the `~/mima_tutorial/output/QC_module`
- Submit the PBS job using `qsub`

```Shell
cd ~/mima_tutorial/output/QC_module
qsub qcModule_0.pbs
```

- You can check the job has been submitted and their status with `qstat`
- Repeat the `qsub` command for each of the `qcModule_1.pbs` and `qcModule_2.pbs` files
- Wait until all PBS jobs have completed

{{% alert color=warning title="PBS log files" %}}
- Usually a PBS job will generate a log file in the same directory from where the job was submitted
- As such we changed directory to where the PBS scripts are saved to submit the jobs because the log files are required for the [Step 5. Generating the QC report.](#step-5-optional-generate-qc-report)
{{% /alert %}}

## Step 4. Check QC outputs

- After **all** PBS job completes, you should have the following outputs

```Shell
tree ~/mima_tutorial/output/QC_module
```

- We only show outputs for one sample below with `...` meaning *"and others"*
- You'll have a set of output files for each sample listed in the `manifest.csv` (provided the corresponding *.fastq files exists)

```Text
 ~/mima_tutorial/output/QC_module
├── CleanReads
│   ├── SRR17380209_clean_1.fq.gz
│   ├── SRR17380209_clean_2.fq.gz
│   └── ...
├── QC_module_0.o3492023
├── qcModule_0.pbs
├── qcModule_1.pbs
├── qcModule_2.pbs
├── QCReport
│   ├── SRR17380209.json
│   ├── SRR17380209.outreport.html
│   └── ...
├── ...
├── SRR17380209_qc_module.log
├── SRR17380209.sh
├── SRR17380209_singletons.fq.gz
└── ...
```

## Step 5. (optional) Generate QC report

- You can generate a summary QC Report after all samples have been quality checked
- This step can be run directly from the command line

```
$ apptainer run --app mima-qc-report $SANDBOX \
-i ~/mima_tutorial/output/QC_module \
-o ~/mima_tutorial/output \
--manifest ~/mima_tutorial/manifest.csv
```

{{% alert color=warning title="Troubleshoot" %}}
If your sequence read files are not saved in your home directory, check you have set up the APPTAINER_BIND environment variable when [binding paths]({{< ref "what-is-container.md#path-binding" >}}).
{{% /alert %}}


The output is a comma separated text file located in `~/mima_tutorial/output/QC_module/QC_report.csv`. You can inspect the first few lines using the `head` command:
```Shell
head ~/mima_tutorial/output/QC_module/QC_report.csv
```

```Text
SampleId,Rawreads_seqs,Derep_seqs,PCR_duplicates(%),Post_QC_seqs,low_quality_reads(%),Host_seqs,Host(%),Clean_reads
SRR17380209,14072002,14045392,0.0809091,9568710,31.87295876113675,482,0.005037251625349707,9568228
SRR17380232,11822934,11713130,0.0706387,11102358,5.214421764293575,62,0.0005584399278063273,11102296
SRR17380236,11756456,11656800,0.0688868,10846582,6.950603939331549,48,0.0004425357223132596,10846534
SRR17380231,12223354,12104874,0.069757,11414164,5.7060486544510916,16,0.00014017671377421947,11414148
SRR17380218,14913690,14874174,0.0856518,11505850,22.645452446636703,256,0.002224955131520053,11505594
SRR17380222,16927002,16905928,0.0980011,11692516,30.83777477344042,320,0.0027367933471290525,11692196
SRR17380118,40510808,40299722,0.235332,40148912,0.3742209437573788,324,0.0008069957163471828,40148588
SRR17380115,40393998,40186880,0.234584,40048860,0.3434454229838196,194,0.00048440829526733096,40048666
SRR17380122,43683114,41809064,0.24366,41697506,0.2668273080688915,1548,0.003712452250741327,41695958
```

{{% alert color="warning" title="Troubleshoot" %}}
Beware that if you have failed PBS log files (*.o) in your input directory, the QC report module may fail. Examine the error messages.
{{% /alert %}}


---

**Next:** you can now proceed with either [Taxonomy Profiling with MIMA]({{< ref "mima-apptainer-taxonomy" >}}) or [Function Profiling with MIMA]({{< ref "mima-apptainer-function" >}}).